incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java

```diff
@@ -15,6 +15,4122 @@
+
+                progress = client.onDecodeSuback(traceId, authorization, buffer, progress, limit, suback);
+                client.decoder = decodePacketType;
+            }
+
+            if (reasonCode != SUCCESS)
+            {
+                client.onDecodeError(traceId, authorization, reasonCode);
+                client.decoder = decodeIgnoreAll;
+            }
+        }
+
+        return progress;
+    }
+
+    private int decodeUnsuback(
+        MqttClient client,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final int length = limit - offset;
+
+        int progress = offset;
+
+        if (length > 0)
+        {
+            int reasonCode = SUCCESS;
+
+            final MqttUnsubackFW unsuback = mqttUnsubackRO.tryWrap(buffer, offset, limit);
+            decode:
+            {
+                if (unsuback == null)
+                {
+                    reasonCode = PROTOCOL_ERROR;
+                    break decode;
+                }
+                else if ((unsuback.typeAndFlags() & 0b1111_1111) != UNSUBACK_FIXED_HEADER)
+                {
+                    reasonCode = MALFORMED_PACKET;
+                    break decode;
+                }
+
+                progress = client.onDecodeUnsuback(traceId, authorization, buffer, progress, limit, unsuback);
+                client.decoder = decodePacketType;
+            }
+
+            if (reasonCode != SUCCESS)
+            {
+                client.onDecodeError(traceId, authorization, reasonCode);
+                client.decoder = decodeIgnoreAll;
+            }
+        }
+
+        return progress;
+    }
+
+    private int decodePublish(
+        MqttClient client,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final int length = limit - offset;
+
+        int progress = offset;
+
+        decode:
+        if (length >= client.decodeablePacketBytes)
+        {
+            int reasonCode = SUCCESS;
+            final MqttPublishFW publish = mqttPublishRO.tryWrap(buffer, offset, offset + client.decodeablePacketBytes);
+
+            final MqttPublishHeader mqttPublishHeader = mqttPublishHeaderRO.reset();
+
+            if (publish == null)
+            {
+                reasonCode = PROTOCOL_ERROR;
+            }
+            else
+            {
+                reasonCode = mqttPublishHeader.decode(client, publish.topicName(), publish.properties(), publish.typeAndFlags());
+            }
+
+            if (reasonCode == SUCCESS)
+            {
+                final int qos = mqttPublishHeader.qos;
+                MqttSubscribeStream subscriber = client.subscribeStreams.get(qos);
+
+                if (!client.existStreamForTopic(mqttPublishHeader.topic))
```

Comment:
This check feels heavy for every message received at the mqtt client.
The topic has been decoded via UTF-8 into String and the matcher is reconstructed on every call.
Can we use subscription id to help make this more efficient for the typical case?